### PR TITLE
Fix broken ETW events

### PR DESCRIPTION
**Pull request by sywhang**
This should fix https://github.com/dotnet/coreclr/issues/25342

When we added `DOTNET_TRACE_CONTEXT`, we were doing a copy-by-value cloning of the existing `MCGEN_TRACE_CONTEXT`, which gets updated by mc.exe - generated callback functions that are called by ETW directly - this caused the copy inside `DOTNET_TRACE_CONTEXT`s to be not properly updated by those mc-generated callbacks (since they have no context on these clones) so when we were using the copied instances to check whether an event is enabled, nothing was enabled. 

Fixing it by keeping the pointer to the `MCGEN_TRACE_CONTEXT`s inside `DOTNET_TRACE_CONTEXT`. 

We might want to do a similar thing for EventPipe provider contexts in the future to keep things consistent but this PR only addresses ETW for now. 

cc @tommcdon

### CODE REVIEW

File: src/inc/eventtracebase.h [hunk 1]
```diff
@@ -79,14 +79,14 @@ enum EtwThreadFlags
 // if the fields in the event are not cheap to calculate
 //
 #define ETW_EVENT_ENABLED(Context, EventDescriptor) \
-    ((MCGEN_ENABLE_CHECK(Context.EtwProvider, EventDescriptor)) || EventPipeHelper::IsEnabled(Context, EventDescriptor.Level, EventDescriptor.Keyword))
+    ((Context.EtwProvider->IsEnabled && McGenEventXplatEnabled(Context.EtwProvider, &EventDescriptor)) || EventPipeHelper::IsEnabled(Context, EventDescriptor.Level, EventDescriptor.Keyword))
```
Comment (tommcdon) [inline src/inc/eventtracebase.h]: Can DOTNET_TRACE_CONTEXT::EtwProvider ever be null?

Comment (sywhang) [reply src/inc/eventtracebase.h]: Nope. It is always assigned: https://github.com/sywhang/coreclr/blob/temp-2/clretwallmain-win.h#L5704

**Comment by tommcdon** [comment]
> We might want to do a similar thing for EventPipe provider contexts in the future to keep things consistent but this PR only addresses ETW for now.

Is the reason purely for consistency or is there also a functional/correctness reason to do the same for EventPipe?

**Comment by sywhang** [comment]
For consistency -- We own the callback sites / methods that update the provider context for EventPipe (as in they're not autogenerated) so those were all modified accordingly when https://github.com/dotnet/coreclr/pull/25221 went in.

**Pull request merged**